### PR TITLE
Add ruff rules UP for pyupgrade

### PR DESCRIPTION
**Pull request by cclauss**
https://docs.astral.sh/ruff/rules/#pyupgrade-up

% `ruff check --select=UP --statistics | sort -k2`
```
 2	UP008	[*] super-call-with-parameters
 1	UP009	[*] utf8-encoding-declaration
 1	UP015	[*] redundant-open-modes
 2	UP027	[*] unpacked-list-comprehension
19	UP030	[*] format-literals
17	UP032	[*] f-string
```
% `ruff check --select=UP --fix --unsafe-fixes`
```
Found 28 errors (21 fixed, 7 remaining).
```
% `ruff rule UP032`
# f-string (UP032)

Derived from the **pyupgrade** linter.

Fix is sometimes available.

## What it does
Checks for `str.format` calls that can be replaced with f-strings.

## Why is this bad?
f-strings are more readable and generally preferred over `str.format`
calls.

## Example
```python
"{}".format(foo)
```

Use instead:
```python
f"{foo}"
```

## References
- [Python documentation: f-strings](https://docs.python.org/3/reference/lexical_analysis.html#f-strings)

### CODE REVIEW

File: icepyx/core/APIformatting.py [hunk 1]
```diff
@@ -84,11 +84,11 @@ def _fmt_readable_granules(dset, **kwds):
...
                 # and ATLAS granule region number
                 if dset in ("ATL07", "ATL10", "ATL20", "ATL21"):
-                    granule_name = "{0}-??_{1}_{2}{3}??_*".format(dset, 14 * "?", t, c)
+                    granule_name = "{}-??_{}_{}{}??_*".format(dset, 14 * "?", t, c)
                 elif dset in ("ATL11",):
-                    granule_name = "{0}_{1}??_*".format(dset, t)
+                    granule_name = f"{dset}_{t}??_*"
                 else:
-                    granule_name = "{0}_{1}_{2}{3}??_*".format(dset, 14 * "?", t, c)
+                    granule_name = "{}_{}_{}{}??_*".format(dset, 14 * "?", t, c)
```
Comment (mfisher87) [inline icepyx/core/APIformatting.py:91]: Why was only 1/3 of these converted to f-string? :thinking:

File: icepyx/core/query.py [hunk 2]
```diff
@@ -136,12 +136,7 @@ def __init__(
             self._temporal = tp.Temporal(date_range, start_time, end_time)
 
     def __str__(self):
-        str = "Extent type: {0} \nCoordinates: {1}\nDate range: ({2}, {3})".format(
-            self._spatial._ext_type,
-            self._spatial._spatial_ext,
-            self._temporal._start,
-            self._temporal._end,
-        )
+        str = f"Extent type: {self._spatial._ext_type} \nCoordinates: {self._spatial._spatial_ext}\nDate range: ({self._temporal._start}, {self._temporal._end})"
```
Comment (mfisher87) [inline icepyx/core/query.py:139]: I personally find this a bit less readable. What about:
```suggestion
        str = (
            f"Extent type: {self._spatial._ext_type}"
            f"\nCoordinates: {self._spatial._spatial_ext}"
            f"\nDate range: ({self._temporal._start}, {self._temporal._end})"
        )
```

File: icepyx/core/query.py [hunk 3]
```diff
@@ -442,9 +437,7 @@ def __init__(
     # Properties
 
     def __str__(self):
-        str = "Product {2} v{3}\n{0}\nDate range {1}".format(
-            self.spatial_extent, self.dates, self.product, self.product_version
-        )
+        str = f"Product {self.product} v{self.product_version}\n{self.spatial_extent}\nDate range {self.dates}"
```
Comment (mfisher87) [inline icepyx/core/query.py:440]: Similar comment about readability!

File: icepyx/quest/dataset_scripts/argo.py [hunk 4]
```diff
@@ -56,9 +56,9 @@ def __str__(self):
             df = "\n" + str(self.argodata.head())
         s = (
             "---Argo---\n"
-            "Parameters: {0}\n"
-            "Pressure range: {1}\n"
-            "Dataframe head: {2}".format(self.params, prange, df)
+            f"Parameters: {self.params}\n"
+            f"Pressure range: {prange}\n"
+            f"Dataframe head: {df}"
```
Comment (mfisher87) [inline icepyx/quest/dataset_scripts/argo.py:61]: So much more readable. The implicit concatenation combined with `.format()` just weirds me out :)

File: icepyx/quest/dataset_scripts/argo.py [hunk 5]
```diff
@@ -363,7 +361,7 @@ def _download_profile(
 
         # Consider any status other than 2xx an error
         if not resp.status_code // 100 == 2:
```
Comment (mfisher87) [inline icepyx/quest/dataset_scripts/argo.py:363]: Thinking out loud, this just jumped out at me, but it's not part of the PR. I'd like to try to use `resp.ok` instead of this pattern eventually.

File: icepyx/core/query.py (hunk 2 above)
Comment (cclauss) [reply icepyx/core/query.py:139]: Funny that a repo that fails to set a ruff `line-length` wants to wrap lines.

File: icepyx/quest/dataset_scripts/argo.py (hunk 5 above)
Comment (cclauss) [reply icepyx/quest/dataset_scripts/argo.py:363]: This `if` covers 200, 201, 202, 299... -- https://requests.readthedocs.io/en/latest/api/#status-code-lookup

Comment (mfisher87) [reply icepyx/quest/dataset_scripts/argo.py:363]: Oh, yeah, good point! Thanks.

File: icepyx/core/APIformatting.py (hunk 1 above)
Comment (cclauss) [reply icepyx/core/APIformatting.py:91]: If the conversion would make the line more difficult to read then the conversion will not be made.

File: icepyx/core/query.py (hunk 2 above)
Comment (cclauss) [reply icepyx/core/query.py:139]: `E501` is the ruff rule that enables `line-length` for the linter.  It is universally ignored in this repo.

File: icepyx/core/APIformatting.py (hunk 1 above)
Comment (mfisher87) [reply icepyx/core/APIformatting.py:91]: Personally, I think it'd be easier to read this one as an f-string! Just my opinion though :) I'm not requesting any changes though, just curious about why it did what it did.

```python
"{}_{}_{}{}??_*".format(dset, 14 * "?", t, c)
f"{dset}_{14 * '?'}_{t}{c}??_*"
f"{dset}_??????????????_{t}{c}??_*"  # I think I like this best haha
```

File: icepyx/core/query.py (hunk 2 above)
Comment (mfisher87) [reply icepyx/core/query.py:139]: That's not quite a complete description of the situation. `E501` is the "line too long rule", true. However, Ruff is a linter and formatter, and uses `line-length` for both, but `E501` only affects linting behavior. The reason why E501 was disabled is explained in the config file in a co-located comment. Please try for yourself! If you add a long line, the formatter will work as described above! This is the formatting that was applied for me:

```
-print("This is a very very very very very very very very very very very long line" + "And I want it to be formatted")
+print(
+    "This is a very very very very very very very very very very very long line"
+    + "And I want it to be formatted"
+)
```

File: icepyx/core/APIformatting.py (hunk 1 above)
Comment (JessicaS11) [reply icepyx/core/APIformatting.py:91]: > If the conversion would make the line more difficult to read then the conversion will not be made.

Any idea what these criteria are? I'm with Matt that I think the f-strings are clearer.

File: icepyx/core/granules.py [hunk 6]
```diff
@@ -498,7 +498,7 @@ def download(self, verbose, path, restart=False):
         # (could create confusion depending on whether download was interrupted or kernel restarted)
         order_fn = ".order_restart"
         if os.path.exists(order_fn):
-            with open(order_fn, "r") as fid:
+            with open(order_fn) as fid:
```
Comment (JessicaS11) [inline icepyx/core/granules.py:501]: Why is being explicit not the preference here (even if "r" is default)?

Comment (mfisher87) [reply icepyx/core/granules.py:501]: I think it's considered well-known or intuitive that read mode is the default? I don't think I fully agree with this. "explicit is better than implicit" seems like a good enough reason to disable this rule, but this is pretty low-stakes IMO, so I don't feel strongly.

The rule docs (https://docs.astral.sh/ruff/rules/redundant-open-modes/) say:

> ## Why is this bad?
> Redundant `open` mode parameters are unnecessary and should be removed to avoid confusion.

Whenever I see "it's unnecessary" used as a justification, it's a yellow-flag in my mind. We do lots of things that are unnecessary to make our work more accessible or provide other value.

**Comment by mfisher87** [comment]
@cclauss  I'm still interested in this PR, are you open to continuing?

**Comment by mfisher87** [comment]
Related conversation happening here: https://github.com/icesat2py/icepyx/pull/572#issuecomment-2301441004
